### PR TITLE
labels(engine): implement pkg/label

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #6270 

### What is changed and how it works?
- Implemented a general purpose mechanism to support labels and selectors.

#### Tests 

 - Unit test

##### Will it cause performance regression or break compatibility?
No.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None.
```

### CODE REVIEW

File: pkg/label/selector.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			log.Warn("illegal regular expression",
+				zap.Any("selector", s))
+
+			// We only make the selector not match,
+			// because it only complicates things to force the caller to
+			// handle such errors that are preventable with a pre-check.
+			return false
+		}
+
+		return regex.MatchString(string(value))
```
Comment (gozssky) [inline pkg/label/selector.go:107]: Do we need to check `exists` here?

File: pkg/label/label.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	return Value(str), true
+}
+
+func isLabelValid(str string) bool {
+	if len(str) == 0 {
+		return false
+	}
+
+	for i := 0; i < len(str); i++ {
+		if !isLabelCharLegal(str[i]) {
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply pkg/label/label.go:55]: Seems k8s label has similar rule, we can also use a regex to match.

https://github.com/kubernetes/kubernetes/blob/09e8339ae4e575ec40e8ed0fc2829fcb678c0f19/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L167

**Review by amyangfei — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

**Pull request merged**